packages/notebook/src/widget.ts

```diff
@@ -892,6 +900,44 @@ export class StaticNotebook extends WindowedList {
     }
   }
 
+  private async _updateForDeferMode(
+    cell: Cell<ICellModel>,
+    cellIdx: number
+  ): Promise<void> {
+    cell.dataset.windowedListIndex = `${cellIdx}`;
+    this.layout.insertWidget(cellIdx, cell);
+    await cell.ready;
+  }
+
+  private _renderCSSAndJSOutputs(
+    cell: Cell<ICellModel>,
+    cellIdx: number
+  ): void {
+    // Only render cell with text/html outputs containing scripts or/and styles
+    // Note:
+    //   We don't need to render JavaScript mimetype outputs because they get
+    //   directly evaluate without adding DOM elements (see @jupyterlab/javascript-extension)
+    if (cell instanceof CodeCell) {
+      for (
+        let outputIdx = 0;
+        outputIdx < cell.model.outputs.length;
+        outputIdx++
+      ) {
+        const output = cell.model.outputs.get(outputIdx);
+        const html =
+          (output.data[htmlRendererFactory.mimeTypes[0]] as string) ?? '';
+        if (
+          html.match(
+            /(<style[^>]*>[^<]*<\/style>|<script[^>]*>.*?<\/script>)/gms
+          )
```

Comment:
So this will always be a hard problem with false positives and negatives. False positives are things like commented out `<script>`:

```html
<!--- <script> </script> --->
```

And false negatives:

```html
<img src="doesnotexist" onerror="alert(1)">
````

I think that this is fine to keep it as-is for transitional period and fail on edge cases (false positive cost is a cell rendered too early, false negative is some very advanced users getting slightly upset but those will know how to workaround it).

However, in the bigger picture, supporting this pattern goes slightly against the goal of introducing style isolation and JS isolation for outputs (shadow DOM). Here is what I would propose:
- we always render cells with `javaScriptRendererFactory` outputs
- we create a new rendered `cssRendererFactory` (name could be different) and always render cells with it too
- we keep rendering cells with HTML if those meet the regex condition **but** emit a future warning in the console (telling users to switch to js/css outputs for notebook-wide effects, or ignore the warning if they intend to decorate HTML content in this output) and this will enable us by JupyterLab 5 or 6 to roll out shadow DOM isolation for HTML outputs (with JS and CSS outputs being rendered as-is).

I do not feel strongly about the plan above, if you don't immediately agree we can hoist it to a new issue. I only feel strongly that we should include `javaScriptRendererFactory`.

Reply:
> So this will always be a hard problem with false positives and negatives.

Yep - there is space to improve this as I'm not checking the _preferred_ output that could be different than `text/html` for instance.

> JavaScript factory

The JavaScript case is already working - see above comment.

> `cssRendererFactory` 

You mean that we should introduce a `%%css` magic or that the cell output with mimetype `text/css` should be rendered?

> Warning

I like the idea; especially as the notebook with contaminent outputs is likely not working as expected on other applications like Colab, VS Code,... .